pcsx2/DEV9/AdapterUtils.cpp

```diff
@@ -244,30 +245,88 @@ std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
 
 	return std::nullopt;
 }
-#elif defined(__POSIX__)
-#ifdef __linux__
+#else
 std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
 {
-	struct ifreq ifr;
-	strcpy(ifr.ifr_name, adapter->ifa_name);
-
-	int fd = socket(AF_INET, SOCK_DGRAM, 0);
-	int ret = ioctl(fd, SIOCGIFHWADDR, &ifr);
-	close(fd);
+	int sd;
+	struct ifreq ifr, *ifrp;
+	struct ifconf ifc;
+	char buf[1024];
+	int n, i;
+	unsigned char *addr;
+#ifdef AF_LINK
+	struct sockaddr_dl *sdlp;
+#endif
 
-	if (ret == 0)
-		return *(MAC_Address*)ifr.ifr_hwaddr.sa_data;
+/*
+ * BSD 4.4 defines the size of an ifreq to be
+ * max(sizeof(ifreq), sizeof(ifreq.ifr_name)+ifreq.ifr_addr.sa_len
+ * However, under earlier systems, sa_len isn't present, so the size is 
+ * just sizeof(struct ifreq)
+ */
+#ifdef HAVE_SA_LEN
+#ifndef max
+#define max(a,b) ((a) > (b) ? (a) : (b))
+#endif
+#define ifreq_size(i) max(sizeof(struct ifreq),\
+     sizeof((i).ifr_name)+(i).ifr_addr.sa_len)
+#else
+#define ifreq_size(i) sizeof(struct ifreq)
+#endif /* HAVE_SA_LEN*/
 
-	return std::nullopt;
-}
+	sd = socket(AF_INET, SOCK_DGRAM, IPPROTO_IP);
+	if (sd < 0) {
+		return std::nullopt;
+	}
+	memset(buf, 0, sizeof(buf));
+	ifc.ifc_len = sizeof(buf);
+	ifc.ifc_buf = buf;
+	if (ioctl (sd, SIOCGIFCONF, (char *)&ifc) < 0) {
+		close(sd);
+		return std::nullopt;
+	}
+	n = ifc.ifc_len;
+	for (i = 0; i < n; i+= ifreq_size(*ifrp) ) {
+		ifrp = (struct ifreq *)((char *) ifc.ifc_buf+i);
+		strncpy(ifr.ifr_name, ifrp->ifr_name, IFNAMSIZ);
+#ifdef SIOCGIFHWADDR
+		if (ioctl(sd, SIOCGIFHWADDR, &ifr) < 0)
+			continue;
+		addr = (unsigned char *) &ifr.ifr_hwaddr.sa_data;
 #else
-std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
-{
-	Console.Error("DEV9: Unsupported OS, can't get MAC address");
+#ifdef SIOCGENADDR
+		if (ioctl(sd, SIOCGENADDR, &ifr) < 0)
+			continue;
+		addr = (unsigned char *) ifr.ifr_enaddr;
+#else
+#ifdef AF_LINK
+		sdlp = (struct sockaddr_dl *) &ifrp->ifr_addr;
+		if ((sdlp->sdl_family != AF_LINK) || (sdlp->sdl_alen != 6))
+			continue;
+		addr = (unsigned char *) &sdlp->sdl_data[sdlp->sdl_nlen];
+#else
+		/*
+		 * XXX we don't have a way of getting the hardware
+		 * address
+		 */
+		Console.Error("DEV9: Unsupported OS, can't get MAC address");
+		close(sd);
+		return std::nullopt
+#endif /* AF_LINK */
+#endif /* SIOCGENADDR */
+#endif /* SIOCGIFHWADDR */
+		if (!addr[0] && !addr[1] && !addr[2] && !addr[3] && !addr[4] && !addr[5])
+			continue;
+			
+		// We have a valid MAC address.
+		close(sd);
+		return *(MAC_Address *)addr;
```

Comment:
Please don't type pun like this, it's undefined behaviour. In this case, it's also looking like you're returning the address of a temporary (but AFAICT it'll be "safe", since it'll get copied before the stack frame is destroyed).

Use a `memcpy()` into the `MAC_Address` struct instead. I know the type pun was there before, but we should try to clean up code when it's changed, not keep the mess around.

Reply:
> Please don't type pun like this, it's undefined behaviour.

Didn't realise this, huh.
Dealing with BSD sockets while avoiding undefined behaviour type pun must be a interesting